pkg/reconciler/apis/crdcleanup/crdcleanup_controller.go

```diff
@@ -0,0 +1,345 @@
+		},
+	}
+
+	indexers.AddIfNotPresentOrDie(
+		apiBindingInformer.Informer().GetIndexer(),
+		cache.Indexers{
+			indexers.APIBindingByBoundResourceUID: indexers.IndexAPIBindingByBoundResourceUID,
+		},
+	)
+
+	crdInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+
+			if !ok {
+				return false
+			}
+
+			return logicalcluster.From(crd) == apibinding.ShadowWorkspaceName
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.enqueueCRD(obj)
+			},
+			UpdateFunc: func(_, newObj interface{}) {
+				c.enqueueCRD(newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueCRD(obj)
+			},
+		},
+	})
+
+	apiBindingInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				c.enqueueFromAPIBinding(oldObj, newObj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.enqueueFromAPIBinding(obj, obj)
+			},
+		},
+	)
+
+	return c, nil
+}
+
+// controller deletes bound CRDs when they are no longer in use by any APIBindings
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	getCRD                           func(clusterName logicalcluster.Name, name string) (*apiextensionsv1.CustomResourceDefinition, error)
+	getAPIBindingsByBoundResourceUID func(name string) ([]interface{}, error)
+	listAPIBindings                  func() ([]*apisv1alpha1.APIBinding, error)
+	getResourceUIDs                  func(binding *apisv1alpha1.APIBinding) ([]types.UID, error)
+	deleteCRD                        func(ctx context.Context, name string) error
+}
+
+// enqueueCRD enqueues a CRD.
+func (c *controller) enqueueCRD(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(4).Info("queueing CRD")
+	c.queue.Add(key)
+}
+
+func (c *controller) enqueueFromAPIBinding(oldObj, newObj interface{}) {
+	oldBinding, ok := oldObj.(*apisv1alpha1.APIBinding)
+	if !ok {
+		return
+	}
+
+	newBinding, ok := newObj.(*apisv1alpha1.APIBinding)
+	if !ok {
+		return
+	}
+
+	logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), newBinding)
+
+	// Looking at old and new versions in case a schema gets removed from an APIExport.
+	// In that case, the last APIBinding to have the schema removed will trigger the CRD delete,
+	// but only the old version will have the reference to the schema.
+
+	uidSet := map[string]bool{}
+
+	for _, boundResource := range oldBinding.Status.BoundResources {
+		uidSet[boundResource.Schema.UID] = false
+	}
+
+	for _, boundResource := range newBinding.Status.BoundResources {
+		uidSet[boundResource.Schema.UID] = false
+	}
```

Comment:
From the comment: It looks at old and new versions in case a schema gets removed from an APIExport. In that case, the last APIBinding to have the schema removed will trigger the CRD delete, but only the old version will have the reference to the schema.

Reply:
What Kyle wrote makes sense to me